overlord/notices/notices.go

```diff
@@ -0,0 +1,336 @@
+	st.Unlock()
+	return nm
+}
+
+// Ensure implements StateManager.Ensure. Required by StateEngine, we
+// actually do nothing here.
+func (m *NoticeManager) Ensure() error {
+	return nil
+}
+
+func (m *NoticeManager) StartUp() {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	m.changeListenerID = m.state.AddChangeStatusChangedHandler(m.onChangeEvent)
+}
+
+func (m *NoticeManager) Stop() {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	m.state.RemoveChangeStatusChangedHandler(m.changeListenerID)
+}
+
+func (m *NoticeManager) onChangeEvent(chg *state.Change, old, new state.Status) {
+	m.addNotice(NoticeChangeUpdate, &NoticeOptions{
+		Key: chg.ID(),
+		Data: map[string]string{
+			"old-status": old.String(),
+			"new-status": new.String(),
+		},
+	})
+}
+
+func noticeUniqueKey(noticeType NoticeType, key string) string {
+	return fmt.Sprintf("%s-%s", noticeType, key)
+}
+
+func updateNotice(notice *Notice, opts *NoticeOptions) bool {
+	now := timeNow().UTC()
+	var notify bool
+	if notice.repeatAfter == 0 || now.After(notice.lastRepeated.Add(notice.repeatAfter)) {
+		// Update last repeated time if repeat-after time has elapsed (or is zero)
+		notice.lastRepeated = now
+		notice.lastData = opts.Data
+		notify = true
+	}
+	return notify
+}
+
+func newNotice(noticeType NoticeType, opts *NoticeOptions) *Notice {
+	now := timeNow().UTC()
+	return &Notice{
+		noticeType:    noticeType,
+		key:           opts.Key,
+		lastData:      opts.Data,
+		firstOccurred: now,
+		lastOccurred:  now,
+		lastRepeated:  now,
+		repeatAfter:   opts.repeatAfter,
+		expireAfter:   NoticeExpireAfter,
+	}
+}
+
+func (m *NoticeManager) addNotice(noticeType NoticeType, opts *NoticeOptions) error {
+	if opts == nil {
+		return fmt.Errorf("internal error: notice options must be provided")
+	}
+
+	key := noticeUniqueKey(noticeType, opts.Key)
+	notify := true
+	if notice := m.notices[key]; notice != nil {
+		notify = updateNotice(notice, opts)
+	} else {
+		m.notices[key] = newNotice(noticeType, opts)
+	}
+	if notify {
+		m.notifyWaiters(m.notices[key])
+	}
+	return nil
+}
+
+func (m *NoticeManager) flattenNotices(filter NoticeFilter) []*Notice {
+	now := timeNow().UTC()
+	var notices []*Notice
+	for _, n := range m.notices {
+		if n.expired(now) || !filter.matches(n) {
+			continue
+		}
+		notices = append(notices, n)
+	}
+	return notices
+}
+
+func (m *NoticeManager) noticesMatchingFilter(filter NoticeFilter) []*Notice {
+	notices := m.flattenNotices(filter)
+	sort.Slice(notices, func(i, j int) bool {
+		return notices[i].lastRepeated.Before(notices[j].lastRepeated)
+	})
+	return notices
+}
+
+func (m *NoticeManager) notifyWaiters(notice *Notice) {
+	for _, w := range m.waiters {
+		if w.filter.matches(notice) {
+			w.notify(notice)
+		}
+	}
+}
+
+// addNoticeWaiter adds a notice-waiter with the given filters. Processing
+// notices for this waiter stops when the done channel is closed.
+func (s *NoticeManager) addNoticeWaiter(filter NoticeFilter) (ch chan *Notice, waiterId int) {
```

Comment:
various of these methods should clarify that they expect to be called with the state lock held, otherwise their manipulation of manage state is racy. in fact they are because they are called from places using the state cache mechanism that requires that